src/plugins/ucx/ucx_backend.cpp

```diff
@@ -421,19 +452,27 @@ nixlUcxEngine::nixlUcxEngine (const nixlBackendInitParams* init_params)
     if (custom_params->count("device_list")!=0)
         devs = str_split((*custom_params)["device_list"], ", ");
 
-    uc = new nixlUcxContext(devs, sizeof(nixlUcxIntReq),
-                           _internalRequestInit, _internalRequestFini, NIXL_UCX_MT_WORKER);
-    uw = new nixlUcxWorker(uc);
-    uw->epAddr(n_addr, workerSize);
-    workerAddr = (void*) n_addr;
-
-    uw->regAmCallback(CONN_CHECK, connectionCheckAmCb, this);
-    uw->regAmCallback(DISCONNECT, connectionTermAmCb, this);
-    uw->regAmCallback(NOTIF_STR, notifAmCb, this);
+    uc = std::make_shared<nixlUcxContext>(devs, sizeof(nixlUcxIntReq),
+                                          _internalRequestInit, _internalRequestFini, NIXL_UCX_MT_WORKER,
+                                          init_params->enableProgTh);
+    for (unsigned int i = 0; i < init_params->numWorkers; i++) {
+        uint64_t n_addr;
+        const auto & uw = uws.emplace_back(std::make_unique<nixlUcxWorker>(uc));
+        uw->epAddr(n_addr, workerSize);
```

Comment:
This is a memory leak as epAddr allocates memory for the address.

We want to do it once at the end of the loop. And my suggestion is to do it for the worker[0].
Then in the progress it will be easy to figure out of `progressNotif` needs to be called (all notifs are coming on the worker which address was shared with others)

Reply:
Good catch, thanks!